File: src/cpp/scheduler.cpp
#include "scheduler.h"
#include <fstream>
#include <iostream>
    
bool Scheduler::initializeSimulation() {

    // Test case: 2 CPUs + 2 GPUs

    // Step 1: Initilize processors and tasks

    simulator.createNewProcessors(CPU, 2);
    simulator.createNewProcessors(GPU, 2);
    // It is recommended to sort in case created in a wrong order
    simulator.sortProcessorsByType();

    simulator.initializeStorages();

    std::ifstream input_file;
    std:: cout << "Reading taskset infomation...\n";
    for (int i = 0; i < 5; i++) {
        std::string filename = "/home/hamster/HeterSchedulerSim/test/.taskset_prop/task";
        // TODO: support task num > 10
        filename = filename + (char)(i+48);
        filename = filename + ".txt";
        input_file.open(filename);
        int deadlines;
        input_file >> deadlines;
        std::vector<unsigned int> segments(10);
        for (int j = 0; j < 2*5-1; j++){
            input_file >> segments[j];
        }
        std::vector<ProcessorAffinity_t> processorTypes = {ProcessorAffinity_t::CPU, ProcessorAffinity_t::GPU};
        Task & tmp = simulator.createNewHeterSSTaskWithVector(processorTypes, segments);
        tmp.setTaskRelativeDeadline(deadlines);
        tmp.setTaskPeriod(deadlines);
        tmp.setTaskRTPriority(99-i);
        input_file.close();
    }

    // Step 2: Start simulation
    simulator.setSimulationTimeBound(1200);
    return true;
}

bool Scheduler::startScheduleLoop() {

    while (! simulator.isSimulationCompleted()) {

        if (!simulator.checkTaskRelease()) return false;

        // TODO: Make decisions here
        if (!makeScheduleDecisions())
            std::cerr << "Scheduling decisions are inproper!\n";

        simulator.printSimulatorStates();

        simulator.updateProcessorAndTask();
    }

    return (!simulator.doesTaskMissDeadline());

}

bool Scheduler::makeScheduleDecisions() {
    // Cureent Implementation: based on SCHED_FIFO

    simulator.sortTasksByPriority();

    // First round: check IDLE Processors
    for (ProcessorIndex_t j = 0; j < simulator.queryProcessorCount(); j++) {
        Processor & proc = simulator.getProcessor(j);
        if (proc.queryProcessorState()!=IDLE) continue;
        for (TaskIndex_t i = 0; i < simulator.queryTaskCount(); i++) {
            Segment & readySegmet = simulator.getTask(i).
                                    getFirstReadySegment(proc.queryProcessorType());
            if (!readySegmet.isSegmentMarkedReady()) continue;
            proc.scheduleTaskSpecifiedSegment(simulator.getTask(i), &readySegmet, simulator.queryCurrentTimeStamp());
            break;
        }
    }
    // Next round: check preemptive processors
    for (ProcessorIndex_t j = 0; j < simulator.queryProcessorCount(); j++) {
        Processor & proc = simulator.getProcessor(j);
        if (proc.queryProcessorState()!=BUSY_PREEMPTIVE) continue;
        for (TaskIndex_t i = 0; i < simulator.queryTaskCount(); i++) {
            Task & task = simulator.getTask(i);
            if (task.queryTaskRTPriority() <= 
                proc.getCurrentTask()->queryTaskRTPriority()) continue;
            Segment & readySegmet = task.getFirstReadySegment(proc.queryProcessorType());
            if (!readySegmet.isSegmentMarkedReady()) continue;
            proc.scheduleTaskSpecifiedSegment(simulator.getTask(i), &readySegmet, simulator.queryCurrentTimeStamp());
            break;
        }
    }

    return true;
}

File: src/python/scheduler.py

from client import SimulatorClient

if __name__ == "__main__":
    # default scheduler example
    simulator = SimulatorClient("../../build/main")

    # Step 1: initilization
    CPU = 0; GPU = 7
    simulator.create_processor(CPU, 2)
    simulator.create_processor(GPU, 2)

    # create 5 tasks
    for i in range(5):
        filename = f"/home/hamster/HeterSchedulerSim/test/.taskset_prop/task{i}.txt"
        with open(filename) as taskfile:
            period: int = int(taskfile.readline().strip())
            line = taskfile.readline().strip()
            segs = tuple(map(int, line.split()))
            simulator.create_heter_ss_task(period, 2, (CPU, GPU), segs)
    simulator.set_simulation_timebound(200)
    simulator.start_simulation()
    simulator.print()

    # step 2: scheduling loop
    procAffinity = [0,0,7,7]

    timeStamp = 0
    while (not simulator.is_simulation_completed()):
        #schedule decision here

        procState = simulator.query_processor_states()
        for i in range(4):
            if procState[i][1] != 0: continue
            for j in range(5):
                period, segState = simulator.query_task_state(j)
                for k, seg in enumerate(segState):
                    affinity = seg[0]; isReady = seg[2]; currentProc = seg[1]
                    if (procAffinity[i] != seg[0]): continue
                    if (isReady==1 and currentProc==999999):
                        simulator.schedule_segment_on_processor(i, j, k)
                procState = simulator.query_processor_states()
                if procState[i][1] != 0: break

        simulator.update_processor_and_task()
        simulator.print()

        timeStamp = timeStamp + 1

        a = input("Enter to continue...")

    simulator.quit()

